Splitting out stable subsets from `template-haskell` (`template-haskell-lift` and `template-haskell-quasiquote`)

TeofilC opened the pull request:
This proposal introduces the  `template-haskell-lift` and `template-haskell-quasiquote` packages as stable subsets of the `template-haskell` API. This is the first step in a wider strategy to publish stable packages that cover uses of `template-haskell` reducing ecosystem breakage from this inherently unstable package.

[Rendered](https://github.com/TeofilC/ghc-proposals/blob/wip/th-lift-and-quasiquote/proposals/0000-splitting-out-stable-interfaces-from-th.rst)

jvanbruegge commented:
It's a bit hard to gauge how many users fall in category A, did you do e.g. a survey of hackage to see how many % of users would benefit from this?

TeofilC commented:
> It's a bit hard to gauge how many users fall in category A, did you do e.g. a survey of hackage to see how many % of users would benefit from this?

It's difficult to tell which category they fall into without reading the code. I did this for the top 50 packages on Hackage that use TH (based on reverse dependencies). And from that it's around 20 out of 50 from a quick glance over the spreadsheet I made.
https://docs.google.com/spreadsheets/d/1IpJuk9x6wIoY57VhRlu218bPv8qiqFUXqPfr-zOB16E/edit?usp=sharing

konsumlamm commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,191 @@
+^^^^^^^^^^^^^^^^^^^^^^^
+Our strategy is informed by the classes of usages of ``template-haskell`` found in the ecosystem. We can divide users as follows:
+
+* (A) Quote-and-splice clients: These users use only splices, quotes, ``DeriveLift`` or quasiquotes. These users might not even need to import the ``template-haskell`` library.
+* (B) Syntax-construction clients. These users construct Template Haskell syntax trees either directly through its constructors, or indirectly through the smart-constructors exported by ``Language.Haskell.TH.Lib``.
+* (C) Reification clients. These users, notably various forms of deriving, use reification to interrogate the program. Reification currently returns Template Haskell ASTs.
+* (D) Syntax-analysis clients. Some clients pattern match on Template Haskell syntax tree datatypes.
```
```suggestion
* (\A) Quote-and-splice clients: These users use only splices, quotes, ``DeriveLift`` or quasiquotes. These users might not even need to import the ``template-haskell`` library.
* (\B) Syntax-construction clients. These users construct Template Haskell syntax trees either directly through its constructors, or indirectly through the smart-constructors exported by ``Language.Haskell.TH.Lib``.
* (\C) Reification clients. These users, notably various forms of deriving, use reification to interrogate the program. Reification currently returns Template Haskell ASTs.
* (\D) Syntax-analysis clients. Some clients pattern match on Template Haskell syntax tree datatypes.
```
This should fix the rendering of (A), (B), etc.

konsumlamm commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,191 @@
+
+These diverse usages of the library lead to diverse levels of breakage when a new major version of ``template-haskell`` comes out. We can rank them from (A) with the least breakage to (D) with the most.
+For instance, the ``uuid`` library, which just depends on ``template-haskell`` in order to provide a derived ``Lift`` instance (a type (A) client), in all likelyhood would only need to bump its upper-bound on the library.
+On the other-hand ``th-desugar``, which pattern matches on the entire syntax tree (a type (D) client), would have to make code changes on most releases of the library.
```
```suggestion
On the other hand ``th-desugar``, which pattern matches on the entire syntax tree (a type (D) client), would have to make code changes on most releases of the library.
```

konsumlamm commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,191 @@
+In the future, we aim to continue this strategy, by identifying stable interfaces for these classes of users which aren't tightly coupled to the Template Haskell AST.
+The smart-constructors from the ``Language.Haskell.TH.Lib`` module are a good starting point for type (B) clients. Another idea is to use smart-constructors based on the Haskell2010 AST (See: `GHC#20828 <https://gitlab.haskell.org/ghc/ghc/-/issues/20828>`_).
+For type (C) clients, we can build on the existing `th-abstraction` library, and perhaps expose a refined AST that doesn't need to be as expressive as the surface language.
+Type (D) clients on the otherhand are likely to be difficult to accommodate, since they are inherently tightly coupled to the Template Haskell syntax trees.
```
```suggestion
Type (D) clients on the other hand are likely to be difficult to accommodate, since they are inherently tightly coupled to the Template Haskell syntax trees.
```

konsumlamm commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,191 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Publishing ``template-haskell-lift`` and ``template-haskell-quasiquote`` will be beneficial both for GHC and the ecosystem.
+
+The biggest benefit is that library authors who are just deriving or using ``Lift`` instances or just exposing ``Quasiquoter``s no longer need to depend on the entirety of ``template-haskell``.
```
```suggestion
The biggest benefit is that library authors who are just deriving or using ``Lift`` instances or just exposing ``Quasiquoter``\s no longer need to depend on the entirety of ``template-haskell``.
```
This should fix the rendering of the code blocks.

konsumlamm reviewed (commented):
This is a very well written proposal IMO! I think this would be quite useful for library authors.

sgraf812 commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,192 @@
+
+Effect and Interactions
+-----------------------
+This works towards removing the special case for ``template-haskell`` in (GR1), but on its own it doesn't achieve it.
```
What is (GR1)? Do you mean (A)?

TeofilC replied:
I've clarified this. It's referring to: https://github.com/ghc-proposals/ghc-proposals/blob/01c0d736919a25e2669296aa71dd58368b09ba72/principles.rst?plain=1#L359

sgraf812 reviewed (commented):
Sounds great! Let's do it.

phadej commented:
Few comments:

- I'm in favor of having small interface packages, especially if they would allow to reogranise dependency graph for better. IMHO, that's alone is enough justification. 
- It's very sure that `Lift` will change if there is any work done to fix TTH unsoundess issues. `Lift` interface is not right for typed template haskell.
    - Luckily, but sadly, to my best knowledge we are years away from anything happening.
- What will happen when/if `DeriveLift` learns new stuff, i.e. could derive instance for more types?
   - The proposal doesn't tell how `template-haskell-lift`'s version would change when `DeriveLift` functionality changes.
   - Currently, as `template-haskell` version is bumped every major GHC, it can also be used to version `DeriveLift`. Though this is semantically not right, it's good enough.
   - This is more general issue of how to version GHC extension, or GHC functionality in general: we use `base` and `template-haskell` as proxies. But **NONE** of proposals have yet addressed what will happen if `base` and/or `template-haskell` version will be decoupled from GHC.

TeofilC commented:
> * This is more general issue of how to version GHC extension, or GHC functionality in general: we use `base` and `template-haskell` as proxies

Yes this is a great point. It is something I had to deal with on the MR.
I just ended up using the `__GLASGOW_HASKELL__` macro.

In the future I see users either depending on `__GLASGOW_HASKELL__` if their behaviour depends on the compiler version or `MIN_VERSION_template_haskell_lift` if the behaviour is dependent on the interface. I think that should handle most cases?

(I will shortly add a paragraph about this to the proposal)

phadej commented:
>  I think that should handle most cases?

It won't handle any cases. I want to be able to specify the support window in the `.cabal` file.

TeofilC commented:
> It won't handle any cases. I want to/ should be able to specify the support window in the `.cabal` file.

I think I understand what you mean. Previously you could use `template-haskell` versions to express which version of GHC you were using, and therefore which implementation of `DeriveLift` you are getting. Since `template-haskell-lift` compiles with multiple versions (and a future minor release could expand the window further!), you can no longer depend on that.

Perhaps you could use `impl(ghc >= X.X.X)` conditionals?

Alternatively you could use the version of `ghc-experimental` for this, since that should continue to change with each version of GHC.

phadej commented:
> Perhaps you could use impl(ghc >= X.X.X) conditionals?

No, I cannot. The mechanism should allow Hackage revisions to shrink/widen the support window. In case of metadata mistakes we should be able to correct them (we includes Hackage trustees which cannot go "blacklist the old release and make new one" route). Conditionals cannot be changed, maybe they could be, but someone has to show that it's "safe" and also implement that.

> Alternatively you could use the version of ghc-experimental for this,

Which essentially waters down any benefit of having stable `template-haskell-lift` (or any stable package), as updates are blocked on `ghc-experimental` version tied to GHC.

The main issue is that I want to be able to say "this works now, and should work until something related changes". This is part of motivation for this proposal: `Lift` doesn't change that often. But how to guard that `DeriveLift` doesn't change. GHC version changes way too often.

I have thought of making "small base" package re-exporing just the stuff I care about in `base`, which wouldn't change as often (and e.g. not re-export all GHC internal stuff which is there); but that won't work as GHC itself changes, and the goal of not worrying about testing every single GHC release won't be achieved.

phadej commented:
Addition, you can think that "small base" is related to `base` the same way as `template-haskell-lift` is related to `template-haskell`: a (relatively) small and stable subset of the bigger and less stable package.

I don't think either will work until the stability of compiler's (extensions, parser, etc) is addressed. I have to mention whitespace related change, `foo x @ bar = ...` syntax changed. That did break stuff. So far GHC hadn't stated how it will deal with such issues in the future. Until there is possibility that such changes may happen, I will insist of having tight upper bounds on GHC version. Therefore as a library author reinstallable `base`, `template-haskell-lift` etc have no value to me.

TeofilC commented:
> `Lift` doesn't change that often. But how to guard that `DeriveLift` doesn't change. GHC version changes way too often.

I see. I think I can't give a very satisfactory answer to that. I think we just have to trust the GHC proposal process that a proposal that would break your code wouldn't be accepted, since that would go against https://github.com/ghc-proposals/ghc-proposals/blob/master/principles.rst#33stability-gr1

phadej commented:
> 3.4   Exceptions (GR2)
> General rule (GR2). We may break (GR1), but only with a compelling reason, and with careful consideration of impact.
> 
> Compelling reasons include
>
> ...
>  To fix a design flaw in the language. For this, a better path is usually to fix the design flaw with a new language extension, embodied in a new language edition. Only exceptionally would we fix a design flaw in a way that breaks programs compiled with existing language editions.

Arguably the `@` whitespace change falls into that category, so I can expect such changes to happen in the future.

Thus I don't think I can not not test each major GHC version, and have some kind of upper bound (via `base` or `template-haskell` proxy or some other means).

That said, I feel that all work like `template-haskell-lift` is wasted, as the "foundational bits" are shaky, not truly stable.

I'm somewhat fine with testing my libraries against each major GHC version, if in return I'd actually get things like cleanup of old warts. The current pseudo-stability "we don't change anything until we do" is just bad. I cannot build on top of that.

maralorn commented:
@phadej Are you saying that you think it is wrong to have the sentence

> Only exceptionally would we fix a design flaw in a way that breaks programs compiled with existing language editions.

in our stability rules or are you saying that we declare too much cases exceptional?

Because I think the committee is *very* committed to the stability rules these days and I would like to see examples after them being written down where we didn’t adhere to them.

TeofilC commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,203 @@
+
+Implementation Plan
+-------------------
+Teo Camarasu has implemented a `MR <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13569>`_
```
```suggestion
Teo Camarasu has implemented an `MR <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13569>`_
```

adamgundry commented:
Thanks for driving this forward @TeofilC! I think this is a good direction of travel.

@phadej raises some valid concerns about stability challenges more broadly, but I don't see them as obstacles to this proposal:
 * There are various different sources of friction on GHC upgrades, e.g. language changes, library changes, bounds bumps. We can't solve them all at once. It's good to make incremental improvements like this (reducing the need for bounds bumps in some common, simple cases), even if that doesn't eliminate all sources of upgrade friction.
 * If the `Lift` interface and/or `DeriveLift` really needs to change, nothing stops `template-haskell-lift` having a new (major) version to accommodate it. It is useful for `template-haskell-lift` to avoid version bumps on most GHC releases, even if it doesn't avoid them every time.
 * I can see the need for Cabal libraries to be able to write constraints on the GHC version in use, but that needs to be resolved separately at the Cabal level, not here. (Perhaps someone could publish a `ghc-version` package series that has no exported modules but depends on a suitable non-reinstallable package for each GHC version?)

tomjaguarpaw commented:
> Perhaps someone could publish a ghc-version package series that has no exported modules but depends on a suitable non-reinstallable package for each GHC version?

I like this idea and I've often wondered why something like this doesn't already exist. It's a hack, but a rather elegant hack nonetheless!

phadej commented:
@adamgundry 

> If the Lift interface and/or DeriveLift really needs to change, nothing stops template-haskell-lift having a new (major) version to accommodate it.

That contract should be made before hand for people being able to depend on it. E.g. in this proposal.

TeofilC commented:
@phadej could you expand on what sort of contract you'd like?

If the `Lift` interface changes then we'd have a new major release, but that sounds obvious to me, but I'm happy to make it explicit.

In terms of DeriveLift, you mentioned the possibility of allowing more types to have derived instances, but I'm not sure if that would warrant a new major release. That's more of a GHC concern, ie, not a property of this library which should be as decoupled as possible.

phadej commented:
> In terms of DeriveLift, you mentioned the possibility of allowing more types to have derived instances, but I'm not sure if that would warrant a new major release. That's more of a GHC concern, ie, not a property of this library which should be as decoupled as possible.

It could warrant a minor release though. So I could use `template-haskell-lift` as a proxy to version `DeriveLift` functionality. In (unlikely but possible) case of `DeriveLift` actually **changing** what it derives, I'd welcome if `template-haskell-lift` had a major release, even though `Lift` doesn't change. TL;DR using `template-haskell-lift` version to version whole `Lift` related functionality, including `DeriveLift` extension, and not only the library part.

If, however, `template-haskell-lift` wouldn't version `DeriveLift`; then the problem of "how to version `DeriveLift`" would be still open. And as consequence, for (my) packages which could depend on `template-haskell-lift` only, I would refrain of doing that, and instead of depend on full `template-haskell` (which I expect to have major versions with each GHC release until further notice).

TeofilC commented:
> It could warrant a minor release though.

I would be happy to commit to always making a minor release if anything related downstream of Lift changes in GHC (and the same for template-haskell-quasiquote). In fact we could always release minor revisions rather than revisions for new GHC releases. Would either of these be enough for your use case?

I can't commit to making a major release if only DeriveLift changes because then we are reintroducing the issues users face with the th library: unnecessary version bumps, coupling with GHC.

phadej commented:
> I can't commit to making a major release if only DeriveLift changes because then we are reintroducing the issues users face with the th library: unnecessary version bumps, coupling with GHC.

**But** users of `template-haskell-lift` are most likely also using `DeriveLift` (and if they aren't then arguably they should).

TeofilC commented:
For instance https://gitlab.haskell.org/ghc/ghc/-/commit/217caad163283c37fb5560188a56511f24b1ee57 changes DeriveLift:

> Derived lift instances will now not generate code with expression
quotations.
>
> Before:
> 
> ```
> data A = A Int deriving Lift
> 
> =>
> lift (A x) = [| A $(lift x) |]
> ```
>
> After:
>
> ```
> lift (A x) = conE 'A `appE` (lift x)
> ```
>

I don't think this change should bump the major version of template-haskell-lift. 

If I committed to versioning DeriveLift then if we had a proper breaking change such that GHC-X accepts a deriving instance that GHC-X+1 doesn't then I couldn't have a version of template-haskell-lift that is compatible with both, and having library versions be compatible with a broad range of GHCs is a key property to for smooth upgrades.

So I would rather version less in this library and suggest folks use other means to version GHC features like DeriveLift, which cannot be decoupled from GHC.

TeofilC commented:
Note as well afaik if there were no other changes to template-haskell in a release and DeriveLift changed then I don't think it would have a major release under the status quo.

Of course template-haskell coincidentally always changes since there's always new additions to the AST

phadej commented:
> If I committed to versioning DeriveLift then if we had a proper breaking change such that GHC-X accepts a deriving instance that GHC-X+1 doesn't then I couldn't have a version of template-haskell-lift that is compatible with both, and having library versions be compatible with a broad range of GHCs is a key property to for smooth upgrades.

If `DeriveLift` changes in such way, I don't think there is a way to do "smooth upgrade". 

> I suggest folks use other means to version GHC features like DeriveLift,

Unfortunate.

Ericson2314 commented:
This is a good proposal, let's do it!

simonpj commented:
I pretty sceptical about using the version number of a *library* as a proxy for versioning *GHC features*.  

Why is `DeriveLift` special?  What about `DeriveTypeable`?  Or `GADTs`?  Or any other language extension. 

And what (precisely) does it mean to say "`DeriveLift` has changed so we need a version bump somewhere"?  I think that spec might be hard to nail down.

TeofilC commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,203 @@
+
+We propose to publish two new libraries: ``template-haskell-lift`` and ``template-haskell-quasiquote``.
+These will be shipped with GHC.
+They will also be buildable from Hackage.
+They will be buildable with at a *minimum* the last 3 versions of GHC.
```
TODO: flesh this out. In particular, we should mention that we will use CPP to shim over different versions of the interface.

TeofilC commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,203 @@
+They will also be buildable from Hackage.
+They will be buildable with at a *minimum* the last 3 versions of GHC.
+The current version of the libraries are compatible with GHC 8.10 and later.
+
```
Mention what are dependencies are.
Mention that the library is reinstallable but shipped with GHC ala containers, bytestring, etc, but not baked in like ghc-internal.

simonpj commented:
A crucial part of the specification is that the new libraries depend only on `ghc-internal` (not `template-haskell`).  That is what unlocks the benefits of Section 1.3.  Let's say that loud and clear.

> They will be buildable with at a minimum the last 3 versions of GHC

I'm not very clear what this means.  It might be an implementation consideration (the version in the repo is managed with CPP so it works for the last three versions of GHC).  But then it would not be part of the *user-facing* specification, which Section 3 is about.  Maybe flesh out the intent here.

phadej commented:
> I pretty sceptical about using the version number of a library as a proxy for versioning GHC features.

And importantly, we do use `base` for this for the lack of better options.

TeofilC commented:
Would the following suggestion from @adamgundry work for you @phadej:

> Perhaps someone could publish a ghc-version package series that has no exported modules but depends on a suitable non-reinstallable package for each GHC version?

I imagine a more fine-grained solution would be better but would that be acceptable for your use-case?

Ericson2314 commented:
@simonpj 

> A crucial part of the specification is that the new libraries depend only on ghc-internal (not template-haskell). That is what unlocks the benefits of Section 1.3. Let's say that loud and clear.

I think what these new libraries is written in terms of is an implementation detail. Everything could move to `ghc-internal`, `template-haskell` itself could become another shim library, and that's still fine.

GHC is free to manage itsef unstable libraries however it likes.

> I'm not very clear what this means. It might be an implementation consideration (the version in the repo is managed with CPP so it works for the last three versions of GHC). But then it would not be part of the user-facing specification, which Section 3 is about. Maybe flesh out the intent here.

Conversely, this is a user facing detail. What it is making is a promise to users that they can do "tick tock" upgrades where the version of these new libraries and the version of GHC *never* needs to be upgraded at the same time. In other words, if you are on and old, EoL, version of both, at least one of:

- Upgrade GHC, then upgrade `template-haskell-*` 
- Upgrade `template-haskell-*`, then upgrade GHC

*must* be available as an option. The effect of this is that smaller atomic upgrades make upgrading easier and safer.

-----

Bottom line is IMO this is is closer to where I want to end up with the `base`-split than we've gotten so far. `base` having a version that goes up with each GHC still means the project is only half-done. With this, however, we've "won", making interfaces that are actually stable across GHC versions. This is fantastic, and I hope everyone will grow to adore it, and then it inspire us to finish the job with `base` also.

TeofilC commented:
@Ericson2314 
> I think what these new libraries is written in terms of is an implementation detail. Everything could move to ghc-internal, template-haskell itself could become another shim library, and that's still fine.

There are two reasons we might not want to depend on `template-haskell`:
1. Wired-in definitions. `template-haskell` used to contain wired-in definitions, but as of [ghc!12479](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12479) this is no longer the case. It is merely a shim over `ghc-internal`. So this reason doesn't apply any more.
2. Breaking up the dependency graph. The other reason is that if we can break up `template-haskell-lift` out of `template-haskell` then we can insert a dependency like `containers` or `filepath` between the two. (`template-haskell` wants to depend on `filepath` but `filepath` needs the quasiquoter and `Lift` interfaces). This is what I outline in Section 1.3.

I think you are referring to (1) but this is justified by (2).
So I should say this explicitly, and also emphasize that there is nothing wired-in about these new packages.

> What it is making is a promise to users that they can do "tick tock" upgrades where the version of these new libraries and the version of GHC never needs to be upgraded at the same time.

Yes I think this is a key detail. I'll add a section about this and illustrate it with a concrete example.

phadej commented:
> Would the following suggestion from @adamgundry work for you @phadej:
> 
> > Perhaps someone could publish a ghc-version package series that has no exported modules but depends on a suitable non-reinstallable package for each GHC version?
> 
> I imagine a more fine-grained solution would be better but would that be acceptable for your use-case?

There is `base`, so `ghc-version` doesn't really add any value *now*.

phadej commented:
Actually, not only `base`, but also `ghc-internal`.

TeofilC commented:
I've added a section that (hopefully) more clearly explains why we want to support a wide range of GHC versions. It includes a concrete example from a previous change to the Lift interface.

I've also clarified that we only depend on `base` and `ghc-internal`, and do *not* depend on `template-haskell`. I'll check if we can drop `base` as a dependency later.

TeofilC commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
````diff
@@ -0,0 +1,254 @@
+Unresolved Questions
+--------------------
+
+- Should the modules live in the ```TemplateHaskell.`` or the ``Language.Haskell.TH.`` namespace?
+- Should these packages live in the GHC repository, in another repository on the GHC Gitlab, or on GitHub?
````
```suggestion
```
I'll drop these before merging

adamgundry commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,254 @@
+They rarely change and if they don't change between two versions of GHC, then we can accommodate both for free.
+If they do change, then it's likely that we can use ``CPP`` to expose to shim over GHC internals and expose a consistent interface.
+
+.. note::
+   For instance, `Overloaded Quotations proposal <./0246-overloaded-bracket.rst>`_ changed the type of the ``lift`` method of ``Lift`` from ``lift :: a -> Q a`` to ``lift :: Qoute m => a -> m a``.
+
+   Suppose ``template-haskell-lift`` existed at the time and ``template-haskell-lift-0.1`` corresponded to the old interface and ``template-haskell-lift-0.2`` corresponded to the new interface.
+   Further suppose that GHC-9.0 ships with ``template-haskell-lift-0.1`` and GHC-9.2 ships with and implements the interface of ``template-haskell-lift-0.2``.
+
+   Our argument in this section is that it is convenient to make the following possible:
+
+   * ``template-haskell-lift-0.1`` can be compiled with GHC-9.2
+   * ``template-haskell-lift-0.2`` can be compiled with GHC-9.0
```
I'm having difficulty following this example, unfortunately. Since `Lift` is a wired-in class, I don't see how `template-haskell-lift` can expose a version of it that uses a different `lift` method type than the one shipped by the compiler. Isn't it merely reexporting whatever `Lift` class the compiler provides?

The situation would be different if the class was exported without methods, of course. I don't know whether `DeriveLift` works if the class is in scope but the methods are not?

(Side note: the RST note rendering doesn't seem to work.)

TeofilC replied:
That's a good point. I should make this a bit clearer.

While we are forced to re-export `Lift` as is from `ghc-internal`, we do have a bit of leeway.

In one direction we export `type Quote m = Q ~ m` (see the [MR](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13569/diffs?file=79e359557bae9fa573412b8b1338fde5f60b608a#79e359557bae9fa573412b8b1338fde5f60b608a_0_34)). I haven't tried out the other direction, but I think we could define `class Quote` and make any instance of `Q` an instance of `Quote`.

So while the types wouldn't be exactly the same, it wouldn't be noticeable to end-users.

We couldn't do this for any arbitrary change, but I think it's likely that we could do it for most.

(I'll integrate this into the text)

Yeah I'm not sure what's going on with the RST note syntax. Perhaps it's just best if I drop it.

TeofilC replied:
On further thought, I think the direction I have on the MR is sound,
but I don't think the other direction is sound, since as you say we can't touch the class method. So an instance could use a method from `Q` not in `Quote`

simonpj commented:
The proposal does not say why we need *two* packages.  Why not just one, for both lift and quasiquote?  Yes, its version has to change if either API changes, but neither will change much anyway.  

At least you should give that as a design alternative, and say why you argue in favour of two, and how strongly

(My personal inclination is to just have one.  Simpler.)

phadej commented:
My view on having more smaller packages:
- Bigger ones are prone to scope creep. Is it `template-haskell-stable-parts`? What if something else in `template-haskell` becomes "stable", should it also be added to this package?
- From my POV both `Lift` and quasiquoters are likely to change.
    - `Lift` will most likely change if TTH soundness issues are fixed.
    - QuasiQuoters don't even support Typed TH, that's my major reason that I'm against using QuasiQuoters: they are "untyped". And the interface is arguably ugly (why I need to do `quasiPat = undefined`).
    
So TL;DR, say I want to improve QuasiQuoters. If I don't need to "poke" `Lift` users for no reason, they'll hopefully appreciate.

Key observation: If `Lift` and `QuasiQuoters` were truly done, users could just depend on `template-haskell` **without having upper bounds** on it!

Ericson2314 commented:
Agree with @phadej 

@simonpj if it helps, let me provide what I think is the "golden rule" of package scope:

> Whenever library/executable a depends on library b, a uses almost all of b.

This prevents version bound churn because of upstream breaking changes that downstream doesn't actually care about. It also corresponds to "every library should do one thing, and one thing well".

------

Also in this specific situation I agree with @phadej that "quasi-quotes" are bad. And per my issue #260, I would in fact deprecate them entirely, now that we have multi-line literals. Even without asking everyone to share these opinions, I think it is clear that someone might just need lifting machinery, and someone else might just needs "quasi-quotes". No need to "force both upon" someone that just needs one.

TeofilC commented:
> The proposal does not say why we need _two_ packages. Why not just one, for both lift and quasiquote? Yes, its version has to change if either API changes, but neither will change much anyway.

@simonpj I've added a section that explains why I don't think this is the right way to go. I've used many of the same arguments as @phadej and @Ericson2314. I think I've covered most of the points but do point out if I'm missing anything.

@adamgundry I've updated the section about updating these libraries independently of GHC with some edits and to clarify the point you raised about both how the `Quote` change wasn't forward compatible.

sgraf812 commented:
Thought experiment: In a world where no Cabal package would ever specify an upper bound on `template-haskell`, stable-subset-packages such as `template-haskell-lift` would provide no additional stability benefits, because packages that only need the `Lift` interface would. (For similar reasons, we split `base` from `ghc-internal`.)

In the world that we live in, the finer-grained the package, the fewer are the APIs that need to be versioned in tandem, the fewer breakage is caused to downstream packages, and the more churn for maintainers. The most radical way to tackle the version bump problem would be to version each symbol individually, specifying (as a haddock comment like `@since`, say) the last breaking change to a certain function.

I think we agree that having `template-haskell-lift` has a good cost/benefit ratio, in that it frees many packages from having to ship a new release/hackage revision whenever a new GHC release happens.   

---

> Why is DeriveLift special? What about DeriveTypeable? Or GADTs? Or any other language extension.

(I think Simon was trying to nudge Teo into adding an explanation to the proposal, but it might be interesting for the discussion to write it out here as well.)

The reason is that the derived `Lift t` implementation essentially returns "serialised" data (TH AST) for the given `t`.
The serialisation protocol may change between GHC versions because the structure of the derived instance may change.
Clients (of type (A), say) usually cannot observe these changes because they observe the serialised data only after interpretation (through splicing), where GHC carefully guarantees that the observed behavior is unchanged, or is very explicit about breaking it.

However, if clients know the definition of `Exp` (which they only can if they depend on `template-haskell` proper!), they may peek into the structure of the serialised data and in doing so make themselves prone to internal changes to the serialisation protocol.
These are not type (A) clients. I think it is a non-goal to help these clients with `template-haskell-lift`.

TeofilC commented:
If I understand correctly a scenario Oleg is worried about is:
1. GHC adds support for DeriveLift for a new class of datatype in GHC-X
2. The ecosystem starts using DeriveLift for that class of datatypes.
3. We realize all too late that there was a fundamental error in the implementation so we have to revert the change in GHC-X+1
4. Now lots of code breaks and the version of `template-haskell-lift` stays the same so we can't use that to tell end-users that they can't use the version of the library

I've picked this since it includes a breaking change forced by a compiler bug, so it's not impacted by our stability principles.

TeofilC commented:
I've added a section about the `DeriveLift` suggestion and why I don't think it should be part of the interface of `template-haskell`

TeofilC commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,335 @@
+
+While we hold these to be valid concerns, we do not think expanding the interface of ``template-haskell-lift`` is the way to resolve them.
+This suggestion goes against our aim of decoupling ``template-haskell-lift`` from GHC.
+It is possible to shim over changes to the ``Lift`` interface in ``ghc-internal`` as we've outlined in :ref:`independence`.
```
This reference works when I actually run sphinx but not on the github preview. I'm not sure if there's a better way to do internal links?

adamgundry replied:
I've tripped over this limitation of GitHub rendering as well, and it's annoying. The only way I know is to link to the automatically-generated anchor, e.g.
```rst
`independence <#131upgrading-libraries-independently-of-ghc>`__
```
and hope the referenced section title doesn't change afterwards.

TeofilC commented:
I think I will wait until after ZuriHac, do another round of edits, and then put this before the committee.

sgraf812 commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,335 @@
+
+We feel that it is reasonable that a library maintainer should be able to explicitly define which versions of GHC they support, but that this can be done orthogonally to this proposal.
+In fact, tackling this independently might lead to a more general solution.
+For instance, we could have a ``ghc-version`` package, which exports nothing but carries the version of GHC. We would have to allow it to be exempted from unused package warnings, but this seems doable.
```
(Non-actionable as far as the proposal goes.) 

Why isn't `ghc-version` simply `ghc-internal`? That is, we could just ensure that `ghc-internal` is versioned according to the GHC version it comes with. Which incidentally is the case *already*. So `ghc-internal` is `ghc-version`.

sgraf812 commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,335 @@
+
+Effect and Interactions
+-----------------------
+This works towards removing the special case for ``template-haskell`` in `(GR1) <https://github.com/ghc-proposals/ghc-proposals/blob/master/principles.rst#33stability-gr1>`_ from `Principles for GHC <../principles.rst>`_, but on its own it doesn't achieve it.
```
(Non-actionable in this proposal.)

I just had a look at (GR1) and don't understand why `template-haskell` warrants a special case at all. Perhaps it used to when it was wired-in, but that's not the case any more since it nowadays re-exports from `ghc-internal`. The versioning of the latter accounts for the alluded breaking changes to GHC's AST.

TeofilC replied:
Yes we should reexamine GR1 and see if we can drop it

sgraf812 reviewed (commented):
Shepherd here. LGTM, except for one small suggestion for the Motivation section. I'll recommend acceptance to the committee.

sgraf812 commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,334 @@
+
+``template-haskell-quasiquote``::
+
+   module TemplateHaskell.Quasiquoter
```
Yes, please make this change (`Q`).

aspiwack commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,335 @@
+`PVP <https://pvp.haskell.org/>`_ dictates that if any interface in a package changes in a breaking way, then the entire package needs to bump its major version.
+The most unstable interface in a package determines the package's overall stability.
+The more interfaces a package contains, the greater the risk of a frequent breaking changes.
+If we choose to continue splitting interfaces out of the ``template-haskell`` package, this risk will grow further if we choose to put them into ``template-haskell-stable``.
+Eventually ``template-haskell-stable`` might no longer be stable at all.
```
The “more splitting means less spurious version bump” is always true. So we could always cut down the packages down to single function packages. I don't think the argument stands on its own. Maybe there's an optimal package size, but I don't know that we have a good way of characterising it in 2025.

The other arguments, I think, are more compelling. Maybe they should be evaluated in lights of the plans to have a better stage distinction (proposal [682](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0682-explicit-level-imports.rst)). @mpickering @alt-romes @adamgundry does the choice of splitting `Lift` into its own package interact at all with your plans? Or are they indifferent to this particular choice?

adamgundry replied:
The argument for separate packages @TeofilC makes in https://github.com/ghc-proposals/ghc-proposals/pull/696#discussion_r2177107282 seems compelling to me. Fundamentally I think `Lift` and quasiquotes are very different features and bundling them together would be more confusing than helpful. Despite the name, quasiquotes aren't (TH) quotes at all, they are splices!

 > The other arguments, I think, are more compelling. Maybe they should be evaluated in lights of the plans to have a better stage distinction (proposal [682](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0682-explicit-level-imports.rst)). @mpickering @alt-romes @adamgundry does the choice of splitting `Lift` into its own package interact at all with your plans? Or are they indifferent to this particular choice?

From the point of view of keeping a strong stage distinction, `Lift` is best avoided where possible, because defining a `Lift` instance inherently forces the datatype involved to be available at both compile-time and runtime (see [section 5.3 of the Explicit Level Imports proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0682-explicit-level-imports.rst#implicit-lifting-and-deriving-lift-instances)). So I think it is potentially valuable for packages to be able to express "I define a quasiquoter, but not any `Lift` instances" via their dependencies.

phadej replied:
> From the point of view of keeping a strong stage distinction, ...

Indeed in library providing the type, the quasiquoter definition would only need the data-type at"compile-time".

However a user of such quasiquoter would actually want to work with the type at "run-time".

So I think that any types "crossing"-through stage persistence will need to exist (i.e. the modules providing them) both at compile and run-time at some point of the pipeline.

TL;DR, I don't think that avoiding `Lift` instances actually saves much. I'd actually think that definining QuasiQuoters using `lift/liftTyped` is a good way to avoid silly mistakes (by writing ad-hoc lift). I.e. I think that library writers using `template-haskell-quasiquoters` should probably depend on `template-haskell-lift` too, but not necessarily vice-versa.

sgraf812 replied:
I think this section is quite long already. Furthermore it is unclear to me what further sentence could be added to improve the argument. 

To me the main point is "From the perspective of the client, the package version number induces a form of tight coupling on the package's API; change in one part of the API affects the shared version number of otherwise loosely coupled different parts of the API. This suggest that every coherent API subset should live in its own package. However, every additional package comes with maintenance overhead. Hence we initiate package splits based on evidence. There are many clients (around xx%) that only use `Lift` and clients that only define quasiquoters; hence these are two different packages to avoid artificial tight coupling.", but apparently YMMV.

sgraf812 commented on proposals/0000-splitting-out-stable-interfaces-from-th.rst:
```diff
@@ -0,0 +1,335 @@
+Our aim here isn't to split these interfaces out in order to fix them in stone, but to make it easier for end-users to cope with future changes,
+and to eliminate unnecessary work when the subset of the interface they depend on hasn't actually changed.
+
+Rather we should be looking at how interfaces are related. We should look at relations of tight coupling.
+If two interfaces are tightly coupled, it makes sense to group them into one package as changes to one will force changes to the other.
```
Ah, actually I wanted to delete this comment in favour of [my response](https://github.com/ghc-proposals/ghc-proposals/pull/696#discussion_r2189155137) on the other thread. I'll delete it.

sgraf812 reviewed (changes requested):
As the shepherd, I'm happy to declare this proposal as accepted, subject to some minor copy editing.

Once done I'll merge this. Good job, @TeofilC! Let's do it.

sgraf812 commented:
Thanks Teo. This proposal has been accepted.

TeofilC commented:
Thanks everyone!
